### DISCUSSION TITLE
Can someone explain me how to use useSerialRequest

### DISCUSSION THREAD

**Discussion by RProduction** [alovajs/alova#181] [General]
Hi, I have some question.
I just don't get how to use useSerialRequest.
I already looked on the documentation, but seem the documentation is lacking in explanation.
I don't get how to create serialHandlers.
In the documentation, to create serialHandler you need to pass a lambda function that return request(args).
First, what is request(args)?
The documentation doesn't show the function of request is pointed to.
The documentation only show request1, request2, request3.
What type is request? Is it requestMethodInstance? can someone give me clearance how can I create serialRequest Hander.

**Comment by JOU-amjs** [comment]
Yes, the requestX function is means that the function returns a `methodInstance`, [detail about methodInstance is here](https://alova.js.org/tutorial/learning/method-instance).

the `useSerialRequest` is a use hook that you can send several requests serially, so each `methodInstance` represents a request.

**Reply by RProduction** [reply]
Alright, I am back.
Thank you for clarification.
I also found it when I fiddle with the code.
I am just going to tell my requirement here.

I actually need to make dynamic request based of array of link. So what I want to do here is to call parallel request of dynamic array of useRequest.

Is there is an elegant way to do that?
In axios you can use axios.all
Can I use alova for that too?

I don't want to use axios back.
Using alova make it easy to place api call in react component with full component lifecycle support.

**Comment by JOU-amjs** [comment]
@starfallproduction According to your situation, you can use `methodInstance` to [send request directly](https://alova.js.org/tutorial/next-step/send-request-directly), and use `Promise.all` to await all responses, example below.

```js
const handleCertainEvent = async () => {
  const promises = [request1(args).send()];
  if (someCondition) {
    promises.push(request2(args).send());
  }
  const [response1, response2] = await Promise.all(promises);
  // ...
}
```

> here requestX is also returns method instance.

this solution need to create and maintain states related request. like `axios`.

In future, it can be considered to write a use hook about parallel, such as `useParallelRequest`😁. now if it is always used in your project, you can consider to cusomize the `useParallelRequest` by yourself via the core use hook of `alova` such as `useRequest/useWatcher/useFetcher`.

**Reply by RProduction** [reply]
It's really helpful if there is hook to handle batch request.

Anyway, I already try it.
I use useEffect and useState to handle the promise.
It's working.
But seem it triggers update and useEffect trigger another request.
The console print HitCache for another request called.

```
Const [boards, setBoards] = useState<Board[]>([]);
Const boardInst = alovaInstanceGenerator(['link']);
Const promises = boardInst.map(v => v.send());

useEffect(() => {
  // fetch batch
}, [promises]);

useEffect(() => {
  // print log boards
}, [boards]);

```

This is my code.

**Reply by JOU-amjs** [reply]
Incorrect usage.
In react, there would be recall the component function when any state is changed, so in your code, it is possible to send requests repeatedly.

my suggestion is that send request in `useEffect`, code below to prevent requesting repeatedly.

```js
Const [boards, setBoards] = useState<Board[]>([]);

useEffect(() => {
  const boardInst = alovaInstanceGenerator(['link']);
  const promises = boardInst.map(v => v.send());
  const res = await Promise.all(promises);
}, []);

useEffect(() => {
  // print log boards
}, [boards]);
```

**Reply by JOU-amjs** [reply]
Just use `alovaInst.Get('xxx').send()` as `axios.get('xxx')`.

**Reply by RProduction** [reply]
Thank you for the code.
At first I use useState to handle the boardInst and promises.
After your suggestion, I refactored the code.
I wrap creation of boardInst and promises inside fetchFunction, so I can pass the function to refresh.

**Comment by RProduction** [comment]
But still, I am still confused about useSerialRequest.
useSerialRequest seem to pass previous request response to next response.
Can you give me an example of use case for useSerialRequest.

**Reply by JOU-amjs** [reply]
I did not provide this example before, I will add one later.

**Reply by JOU-amjs** [reply]
[here](https://alova.js.org/tutorial/example/serial-request) is the example of serial request.😊

**Comment by RProduction** [comment]
Alright thank you very much for helping me.
Let close this discussion

**Comment by JOU-amjs** [comment]
@starfallproduction I'm doing an survey about "Why using alova". Could you please tell me why replace `axios` with `alova`? is `axios` anywhere not good at? or is `alova` anywhere good at than `axios`?

**Reply by RProduction** [reply]
Well actually one of the reason why I decided to use alova is I am trying new library another than axios to handle web request.

Axios is indeed easy to use. But when I try alova, alova seem to focus on implementing web request along side component lifecycle.
It really make me easy to implement web request and handle state easily along side react component. It also reducing code needed to implement API request.

Well it is not like axios will be abandon. I don't to much about alova, but I think alova shine when used in web framework like react and vue.

Axios will be used if i don't use react.
That's all.

Also axios adapter for alova really make my work easy. When I am using global fetch adapter, I need to convert response stream manually.

**Reply by JOU-amjs** [reply]
Very satisfactory answer, thanks
